unittests/test_variables.py

```diff
@@ -0,0 +1,152 @@
+# Copyright 2016-2021 Swiss National Supercomputing Centre (CSCS/ETH Zurich)
+# ReFrame Project Developers. See the top-level LICENSE file for details.
+#
+# SPDX-License-Identifier: BSD-3-Clause
+
+
+import pytest
+
+
+import reframe as rfm
+from reframe.core.fields import Field
+
+
+@pytest.fixture
+def novars():
+    class NoVars(rfm.RegressionTest):
+        pass
+    yield NoVars
+
+
+@pytest.fixture
+def onevar(novars):
+    class OneVar(novars):
+        var('foo', int, value=10)
+    yield OneVar
+
+
+def test_custom_var(onevar):
+    assert not hasattr(onevar, 'foo')
+    inst = onevar()
+    assert hasattr(onevar, 'foo')
+    assert isinstance(onevar.foo, Field)
+    assert hasattr(inst, 'foo')
+    assert inst.foo == 10
+
+
+def test_instantiate_and_inherit(novars):
+    inst = novars()
+    with pytest.raises(ValueError):
+        class MyTest(novars):
+            '''Error from name clashing'''
+
+
+def test_redeclare_var_clash(novars):
+    with pytest.raises(ValueError):
+        class MyTest(novars):
+            var('name', str)
+
+
+def test_inheritance_clash(novars):
+    class MyMixin(rfm.RegressionMixin):
+        var('name', str)
+
+    with pytest.raises(ValueError):
+        class MyTest(novars, MyMixin):
+            '''Trigger error from inheritance clash.'''
+
+
+def test_var_space_clash():
+    class Spam(rfm.RegressionMixin):
+        var('v0', int, value=1)
+
+    class Ham(rfm.RegressionMixin):
+        var('v0', int, value=2)
+
+    with pytest.raises(ValueError):
+        class Eggs(Spam, Ham):
+            '''Trigger error from var name clashing.'''
+
+
+def test_double_declare():
+    with pytest.raises(ValueError):
+        class MyTest(rfm.RegressionTest):
+            var('v0', int, value=1)
+            var('v0', float, value=0.5)
+
+
+def test_double_action_on_var():
+    with pytest.raises(ValueError):
+        class MyTest(rfm.RegressionTest):
+            set_var('v0', 2)
```

Comment:
No, that's checked below by the test `test_set_undef`. This just tests that a double action on a var is not allowed in the same class body.

Reply:
Yes, but my question is if `set_var()` cannot be before `var()`, then you are not sure the error is set because of the double action. Shouldn't it be `var()` first and `set_var()` afterwards?

Reply:
I see your point. I've changed the order.

However, in this case the outcome would be the same regardless of the order. But I do see how can it be confusing from the unit test point of view.